File: packages/KNIRVSDK/npm-package/src/actuarial.ts
/**
 * Browser-safe facade for KNIRVSERVER's actuarial API. The Rust core owns the
 * matching service; this facade is intentionally transport-only for WASM apps.
 */
export class ActuarialClient {
  private headers: HeadersInit = {};
  constructor(private readonly baseUrl = "", private readonly fetcher: typeof fetch = fetch) {}
  setHeaders(headers: HeadersInit) { this.headers = headers; return this; }

  private request<T>(path: string, init?: RequestInit): Promise<T> {
    const headers = new Headers(this.headers);
    headers.set("content-type", "application/json");
    new Headers(init?.headers).forEach((value, name) => headers.set(name, value));
    return this.fetcher(`${this.baseUrl}/api/v1/actuarial${path}`, {
      ...init,
      headers,
    }).then(async response => {
      const body = await response.json();
      if (!response.ok) throw new Error(body?.error ?? `actuarial request failed (${response.status})`);
      return body as T;
    });
  }
  riskClasses<T = unknown>() { return this.request<T>("/risk-classes"); }
  pools<T = unknown>() { return this.request<T>("/pools"); }
  pool<T = unknown>(id: string) { return this.request<T>(`/pools/${encodeURIComponent(id)}`); }
  submission<T = unknown>(id: string) { return this.request<T>(`/submissions/${encodeURIComponent(id)}`); }
  decision<T = unknown>(id: string) { return this.request<T>(`/decisions/${encodeURIComponent(id)}`); }
  settlement<T = unknown>(id: string) { return this.request<T>(`/settlements/${encodeURIComponent(id)}`); }
  createSubmission<T = unknown>(body: unknown) { return this.request<T>("/submissions", { method: "POST", body: JSON.stringify(body) }); }
  createStake<T = unknown>(poolID: string, body: unknown) { return this.request<T>(`/pools/${encodeURIComponent(poolID)}/stakes`, { method: "POST", body: JSON.stringify(body) }); }
  requestStakeExit<T = unknown>(stakeID: string, body: unknown) { return this.request<T>(`/stakes/${encodeURIComponent(stakeID)}/exit`, { method: "POST", body: JSON.stringify(body) }); }
  createSubmissionArtifacts<T = unknown>(submissionID: string, body: unknown) { return this.request<T>(`/submissions/${encodeURIComponent(submissionID)}/artifacts`, { method: "POST", body: JSON.stringify(body) }); }
  requestQuote<T = unknown>(submissionID: string, body: unknown) { return this.request<T>(`/submissions/${encodeURIComponent(submissionID)}/quote`, { method: "POST", body: JSON.stringify(body) }); }
  claimSubmission<T = unknown>(id: string, request: { resolver_wallet: string; signed_intent?: unknown }) { return this.request<T>(`/submissions/${encodeURIComponent(id)}/claim`, { method: "POST", body: JSON.stringify(request) }); }
  createPayoutDestination<T = unknown>(body: unknown) { return this.request<T>("/payout-destinations", { method: "POST", body: JSON.stringify(body) }); }
  createRiskClass<T = unknown>(body: unknown) { return this.request<T>("/risk-classes", { method: "POST", body: JSON.stringify(body) }); }
  createEnterpriseCredential<T = unknown>(body: unknown) { return this.request<T>("/credentials/enterprise", { method: "POST", body: JSON.stringify(body) }); }
  exposure<T = unknown>(wallet?: string) {
    const query = wallet ? `?wallet=${encodeURIComponent(wallet)}` : '';
    return this.request<T>(`/me/exposure${query}`);
  }
  poolReport<T = unknown>(id: string) { return this.request<T>(`/reports/pools/${encodeURIComponent(id)}`); }
}

File: packages/KNIRVSDK/core/src/actuarial.rs
//! Typed access to KNIRVSERVER's unified actuarial syndicate API.
//! Endpoint paths live here so applications do not each grow a private BFF.
//!
//! ## Signing decision
//! `knirv.action.v1` is reserved for Cosmos `SIGN_MODE_DIRECT` transactions.
//! Actuarial requests are authenticated API operations, so their
//! `signed_intent` compatibility field carries the SDK's canonical
//! `knirv.message.v1` `SignedMessage` over its canonical JSON payload. This
//! deliberately avoids treating an API mutation as a chain transaction while
//! retaining one audited SDK signing implementation. New consumers must use
//! `SignedActuarialIntent`, not introduce a second signature format.
use crate::{client::HttpClient, error::Result, types::*};
use serde::Serialize;

#[derive(Clone, Debug)]
pub struct ActuarialService { http: HttpClient }

impl ActuarialService {
    pub fn new(http: HttpClient) -> Self { Self { http } }
    async fn post<T: serde::de::DeserializeOwned, R: Serialize>(&self, path: &str, request: &R) -> Result<T> {
        self.http.post(path, serde_json::to_value(request)?).await
    }
    pub async fn risk_classes(&self) -> Result<Vec<ActuarialRiskClass>> { self.http.get("/api/v1/actuarial/risk-classes", &[]).await }
    pub async fn pools(&self) -> Result<Vec<SyndicatePool>> { self.http.get("/api/v1/actuarial/pools", &[]).await }
    pub async fn pool(&self, id: &str) -> Result<SyndicatePool> { self.http.get(&format!("/api/v1/actuarial/pools/{id}"), &[]).await }
    pub async fn create_pool(&self, request: &CreateSyndicatePoolRequest) -> Result<SyndicatePool> { self.post("/api/v1/actuarial/pools", request).await }
    pub async fn create_stake(&self, pool_id: &str, request: &CreateStakeRequest) -> Result<StakePosition> { self.post(&format!("/api/v1/actuarial/pools/{pool_id}/stakes"), request).await }
    pub async fn request_stake_exit(&self, stake_id: &str, request: &RequestStakeExitRequest) -> Result<()> { self.post(&format!("/api/v1/actuarial/stakes/{stake_id}/exit"), request).await }
    pub async fn submission(&self, id: &str) -> Result<VulnerabilitySubmission> { self.http.get(&format!("/api/v1/actuarial/submissions/{id}"), &[]).await }
    pub async fn submissions(&self, resolver_wallet: Option<&str>, status: Option<&str>) -> Result<Vec<VulnerabilitySubmission>> {
        let mut query = Vec::new();
        if let Some(wallet) = resolver_wallet { query.push(("resolver_wallet", wallet.to_owned())); }
        if let Some(status) = status { query.push(("status", status.to_owned())); }
        self.http.get("/api/v1/actuarial/submissions", &query).await
    }
    pub async fn create_submission(&self, request: &CreateSubmissionRequest) -> Result<VulnerabilitySubmission> { self.post("/api/v1/actuarial/submissions", request).await }
    pub async fn claim_submission(&self, id: &str, request: &ClaimSubmissionRequest) -> Result<VulnerabilitySubmission> { self.post(&format!("/api/v1/actuarial/submissions/{id}/claim"), request).await }
    pub async fn create_submission_artifacts(&self, id: &str, request: &CreateSubmissionArtifactsRequest) -> Result<ValidationResult> { self.post(&format!("/api/v1/actuarial/submissions/{id}/artifacts"), request).await }
    pub async fn request_quote(&self, id: &str, request: &RequestQuoteRequest) -> Result<QuoteResult> { self.post(&format!("/api/v1/actuarial/submissions/{id}/quote"), request).await }
    pub async fn decision(&self, id: &str) -> Result<PricingDecision> { self.http.get(&format!("/api/v1/actuarial/decisions/{id}"), &[]).await }
    pub async fn settlement(&self, id: &str) -> Result<Settlement> { self.http.get(&format!("/api/v1/actuarial/settlements/{id}"), &[]).await }
    pub async fn create_risk_class(&self, request: &CreateActuarialRiskClassRequest) -> Result<ActuarialRiskClass> { self.post("/api/v1/actuarial/risk-classes", request).await }
    pub async fn enterprise_credential(&self, request: &CreateEnterpriseCredentialRequest) -> Result<EnterpriseCredential> { self.post("/api/v1/actuarial/credentials/enterprise", request).await }
    pub async fn credential_challenge(&self, request: &CredentialChallengeRequest) -> Result<CredentialChallenge> { self.post("/api/v1/actuarial/credentials/challenge", request).await }
    pub async fn exposure(&self, wallet: &str) -> Result<ExposureSummary> { self.http.get("/api/v1/actuarial/me/exposure", &[("wallet", wallet.into())]).await }
    pub async fn pool_report(&self, id: &str) -> Result<PoolReport> { self.http.get(&format!("/api/v1/actuarial/reports/pools/{id}"), &[]).await }
    pub async fn create_payout_destination(&self, request: &CreatePayoutDestinationRequest) -> Result<PayoutDestination> { self.post("/api/v1/actuarial/payout-destinations", request).await }
}
